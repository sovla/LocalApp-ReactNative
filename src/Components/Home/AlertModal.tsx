import React from 'react';
import {StyleSheet, Text, View} from 'react-native';

const AlertModal = () => {
    return (
        <View>
            <Text>AlertModal</Text>
        </View>
    );
};

export default AlertModal;

const styles = StyleSheet.create({});
