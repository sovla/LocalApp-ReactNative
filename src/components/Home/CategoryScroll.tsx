import {
  FlatList,
  Image,
  NativeScrollEvent,
  NativeSyntheticEvent,
  ScrollView,
  StyleSheet,
  TouchableOpacity,
  View,
} from 'react-native';
import React, {useCallback, useRef, useState} from 'react';
import {getHeightPixel, getPixel} from '@/Util/pixelChange';
import {categoryMenu} from '@/assets/global/dummy';
import {CategoryCardProps} from '@/Types/Components/HomeTypes';
import {Text} from '../Global/text';
import {useTranslation} from 'react-i18next';
import {useAppNavigation, useAppSelector} from '@/Hooks/CustomHook';
import Theme from '@/assets/global/Theme';
import {onScrollSlide} from '@/Util/Util';

const CategoryScroll = () => {
  const ref = useRef<ScrollView>(null);
  const [dotList, setDotList] = useState<Array<number>>(
    new Array(Math.floor(categoryMenu.length / 5 + 1)).fill(0),
  );
  const [dotNumber, setDotNumber] = useState<number>(0);

  return (
    <View style={styles.scrollContainer}>
      <ScrollView
        ref={ref}
        horizontal
        pagingEnabled
        onMomentumScrollEnd={e => onScrollSlide(e, setDotNumber)}>
        {categoryMenu.map((item, index) => {
          return (
            <>
              <CategoryCard name={item.name} image={item.image} />
              <View
                style={{
                  marginRight: getPixel((index + 1) % 5 === 0 ? 0 : 2), // 카드사이 2px  마진
                }}
              />
            </>
          );
        })}
        <View
          style={{
            width: getPixel((5 - (categoryMenu.length % 5)) * 66),
          }}
        />
      </ScrollView>
      <View style={styles.dotContainer}>
        {dotList.map((item, index) => {
          return <Dot isOn={dotNumber === index} />;
        })}
      </View>
    </View>
  );
};

const CategoryCard: React.FC<CategoryCardProps> = ({name, image}) => {
  const {t} = useTranslation();
  const fontSize = useAppSelector(state => state.fontSize.value);
  const navigation = useAppNavigation();

  const onPressItem = useCallback(() => {
    navigation.navigate('Search', {
      category: name,
    });
  }, []);

  return (
    <TouchableOpacity onPress={onPressItem} style={styles.cardContainer}>
      <View style={styles.cardImageView}>
        <Image source={image} style={styles.cardImage} />
      </View>
      <Text fontSize={`${10 * fontSize}`}>{t(name)}</Text>
    </TouchableOpacity>
  );
};

const Dot: React.FC<{isOn: boolean}> = ({isOn}) => {
  return (
    <View
      style={{
        width: getPixel(7),
        height: getPixel(7),
        borderColor: Theme.color.gray,
        borderWidth: getPixel(1.5),
        borderRadius: 100,
        backgroundColor: isOn ? Theme.color.gray : Theme.color.white,
        marginRight: getPixel(5),
      }}
    />
  );
};
export default CategoryScroll;

const styles = StyleSheet.create({
  dotContainer: {
    marginTop: getHeightPixel(20),
    width: getPixel(328),
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  cardImageView: {
    width: getPixel(46),
    height: getPixel(46),
    marginTop: getHeightPixel(9),
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: Theme.color.white,
    borderRadius: 100,
  },
  cardImage: {width: getPixel(25), height: getPixel(25)},
  cardContainer: {
    width: getPixel(64),
    height: 'auto',
    minHeight: getHeightPixel(83),
    alignItems: 'center',
  },
  scrollContainer: {
    marginTop: getHeightPixel(20),
    marginHorizontal: getPixel(16),
    width: getPixel(328),
  },
});
