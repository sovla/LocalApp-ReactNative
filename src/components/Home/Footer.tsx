import {Image, StyleSheet, TouchableOpacity, View} from 'react-native';
import React, {useState} from 'react';
import {getHeightPixel, getPixel} from '@/Util/pixelChange';
import Theme from '@/assets/global/Theme';
import {Text} from '../Global/text';
import HomeOnIcon from '@/assets/image/home_color.png';
import HomeOffIcon from '@/assets/image/home.png';
import FavoriteOnIcon from '@/assets/image/favorite_color.png';
import FavoriteOffIcon from '@/assets/image/favorite_gray.png';
import ChatOnIcon from '@/assets/image/chat_color.png';
import ChatOffIcon from '@/assets/image/chat_gray.png';
import ProfileOnIcon from '@/assets/image/profile_color.png';
import ProfileOffIcon from '@/assets/image/profile_gray.png';
import {MenuBoxProps} from '@/Types/Components/HomeTypes';
import {useTranslation} from 'react-i18next';
import {useAppSelector} from '@/Hooks/CustomHook';

const Footer = () => {
  const [selectMenu, setSelectMenu] = useState('home');
  return (
    <View style={styles.footerContainer}>
      <MenuBox
        OffImage={HomeOffIcon}
        onImage={HomeOnIcon}
        name="home"
        selectMenu={selectMenu}
        setSelectMenu={setSelectMenu}
      />
      <MenuBox
        OffImage={FavoriteOffIcon}
        onImage={FavoriteOnIcon}
        name="favorite"
        selectMenu={selectMenu}
        setSelectMenu={setSelectMenu}
      />
      <MenuBox
        OffImage={ChatOffIcon}
        onImage={ChatOnIcon}
        name="chat"
        selectMenu={selectMenu}
        setSelectMenu={setSelectMenu}
      />
      <MenuBox
        OffImage={ProfileOffIcon}
        onImage={ProfileOnIcon}
        name="profile"
        selectMenu={selectMenu}
        setSelectMenu={setSelectMenu}
      />
    </View>
  );
};

const MenuBox: React.FC<MenuBoxProps> = ({
  onImage,
  OffImage,
  selectMenu,
  setSelectMenu,
  name,
}) => {
  const {t} = useTranslation();
  const fontSize = useAppSelector(state => state.fontSize.value);
  const onPress = () => {
    setSelectMenu(name);
  };
  return (
    <TouchableOpacity onPress={onPress} style={styles.viewCenter}>
      <Image
        source={selectMenu === name ? onImage : OffImage}
        style={styles.image}
      />
      <Text
        bold={selectMenu === name}
        color={selectMenu === name ? Theme.color.blue : Theme.color.gray}
        fontSize={`${12 * fontSize}`}>
        {t(name)}
      </Text>
    </TouchableOpacity>
  );
};

export default Footer;

const styles = StyleSheet.create({
  image: {
    width: getPixel(20),
    height: getPixel(20),
  },
  viewCenter: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  footerContainer: {
    width: getPixel(360),
    height: getHeightPixel(54),
    paddingHorizontal: getPixel(16),
    backgroundColor: Theme.color.white,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-around',
  },
});
