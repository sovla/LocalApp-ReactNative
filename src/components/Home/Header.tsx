import {
  Image,
  ImageBackground,
  Modal,
  StyleSheet,
  TouchableOpacity,
  View,
} from 'react-native';
import React, {useState} from 'react';

import BackGroundImage from '@assets/image/BG.png';
import {getHeightPixel, getPixel} from '@/Util/pixelChange';
import LocationWhiteIcon from '@assets/image/location_white.png';
import SearchIcon from '@assets/image/search_white.png';
import MenuIcon from '@assets/image/bar_white.png';
import AlarmIcon from '@assets/image/notice_white.png';
import {Box, RowBox} from '../Global/container';
import {WhiteText} from '../Global/text';
import {useAppSelector} from '@/Hooks/CustomHook';
import {useTranslation} from 'react-i18next';
import Theme from '@/assets/global/Theme';
import TrianglePinkIcon from '@assets/image/triangle_pink.png';
import {HeaderProps} from '@/Types/Components/HomeTypes';
import Location from '../Modal/Location';
import useBoolean from '@/Hooks/useBoolean';
import ModalFilter from './ModalFilter';
import ModalMyPage from './ModalMyPage';

const Header: React.FC<HeaderProps> = () => {
  const {t} = useTranslation();
  const fontSize = useAppSelector(state => state.fontSize.value);
  const {
    value: isModal,
    on: onIsModal,
    off: offIsModal,
    toggle: toggleIsModal,
  } = useBoolean(false);
  const {
    value: isMenu,
    on: onIsMenu,
    off: offIsMenu,
    toggle: toggleIsMenu,
  } = useBoolean(false);

  return (
    <ImageBackground style={styles.headerContainer} source={BackGroundImage}>
      <RowBox>
        <Box style={styles.firstIcon}>
          <Image style={[styles.icon]} source={LocationWhiteIcon} />

          <View style={styles.locationCheck}>
            <Image source={TrianglePinkIcon} style={styles.locationCheckIcon} />
            <WhiteText fontSize={`${12 * fontSize}px`}>
              {t('locationCheck')}
            </WhiteText>
          </View>
        </Box>
        <TouchableOpacity onPress={toggleIsModal} style={styles.locationTouch}>
          <WhiteText fontSize={`${18 * fontSize}`}>Bom Retiro</WhiteText>
        </TouchableOpacity>
      </RowBox>
      <RowBox>
        <TouchableOpacity>
          <Image style={styles.icon} source={SearchIcon} />
        </TouchableOpacity>
        <TouchableOpacity onPress={onIsMenu}>
          <Image style={styles.icon} source={MenuIcon} />
        </TouchableOpacity>
        <TouchableOpacity>
          <Image style={styles.icon} source={AlarmIcon} />
        </TouchableOpacity>
      </RowBox>
      {isModal && (
        <Modal
          animationType="slide"
          transparent
          style={{flex: 1, backgroundColor: '#0006'}}
          onRequestClose={offIsModal}>
          <Location offIsModal={offIsModal} />
        </Modal>
      )}
      {isMenu && (
        <Modal visible={isMenu} transparent onRequestClose={offIsMenu}>
          {isMenu && <ModalMyPage onClose={offIsMenu} />}
        </Modal>
      )}
    </ImageBackground>
  );
};

export default Header;

const styles = StyleSheet.create({
  headerContainer: {
    width: '100%',
    height: getHeightPixel(50),
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    zIndex: 100,
  },
  locationCheckIcon: {
    width: getPixel(15),
    height: getHeightPixel(15),
    position: 'absolute',
    top: -getHeightPixel(8),
    left: getPixel(23),
    zIndex: 100,
  },
  locationCheck: {
    position: 'absolute',
    zIndex: 100,
    backgroundColor: Theme.color.pinkRed,
    width: getPixel(120),
    height: getHeightPixel(23),
    bottom: -getHeightPixel(32),
    borderRadius: getPixel(4),
    left: 0,
    justifyContent: 'center',
    alignItems: 'center',
  },
  locationTouch: {flexDirection: 'row', alignItems: 'center'},
  icon: {
    width: getPixel(24),
    height: getPixel(24),
    marginRight: getPixel(10),
  },
  firstIcon: {
    marginLeft: getPixel(16),
  },
});
