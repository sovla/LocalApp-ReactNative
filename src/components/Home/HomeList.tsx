import {FlatList, Image, StyleSheet, View} from 'react-native';
import React from 'react';
import {BoldText, Text} from '../Global/text';
import {useAppSelector} from '@/Hooks/CustomHook';
import {useTranslation} from 'react-i18next';
import {getHeightPixel, getPixel} from '@/Util/pixelChange';
import {RowBox, TouchBox} from '../Global/container';
import MenuListOnIcon from '@assets/image/menu_list_ver.png';
import MenuListOffIcon from '@assets/image/menu_list_ver1.png';
import MenuOffIcon from '@assets/image/menu_ver.png';
import MenuOnIcon from '@assets/image/menu_ver1.png';

const HomeList: React.FC<{
  location: string;
  isList: boolean;
  setIsList: any;
}> = ({location = 'Bom Retiro', isList, setIsList}) => {
  const {t} = useTranslation();
  const fontSize = useAppSelector(state => state.fontSize.value);
  return (
    <View style={styles.mainContainer}>
      <Text fontSize={`${14 * fontSize}px`}>{location + t('HomeTitle')}</Text>
      <View style={styles.titleBetween}>
        <BoldText fontSize={`${20 * fontSize}px`}>{t('HomeTitle1')}</BoldText>
        <RowBox>
          <TouchBox
            onPress={() => {
              setIsList(true);
            }}>
            <Image
              source={isList ? MenuListOnIcon : MenuListOffIcon}
              style={styles.menuImage}
            />
          </TouchBox>
          <TouchBox
            onPress={() => {
              setIsList(false);
            }}>
            <Image
              source={isList ? MenuOffIcon : MenuOnIcon}
              style={[styles.menuImage, styles.menuImageMarginLeft]}
            />
          </TouchBox>
        </RowBox>
      </View>
    </View>
  );
};

export default HomeList;

const styles = StyleSheet.create({
  menuImageMarginLeft: {
    flex: 1,
    marginLeft: getPixel(10),
  },
  menuImage: {width: getPixel(24), height: getPixel(24)},
  mainContainer: {
    marginHorizontal: getPixel(16),
    marginVertical: getHeightPixel(20),
  },
  titleBetween: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    width: '100%',
  },
});
