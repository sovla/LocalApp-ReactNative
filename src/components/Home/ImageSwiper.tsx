import {
  View,
  Text,
  StyleSheet,
  ScrollView,
  Image,
  NativeSyntheticEvent,
  NativeScrollEvent,
} from 'react-native';
import React, {useState} from 'react';
import {ImageSwiperProps} from '@/Types/Components/HomeTypes';
import {getHeightPixel, getPixel} from '@/Util/pixelChange';
import {onScrollSlide} from '@/Util/Util';
import Theme from '@/assets/global/Theme';

export default function ImageSwiper({
  imageArray,
  setImageArray,
  width,
  height,
}: ImageSwiperProps): JSX.Element {
  const [dotNumber, setDotNumber] = useState(0);

  const dotArray = new Array(imageArray?.length).fill(1);

  return (
    <View style={styles.mainContainer}>
      <ScrollView
        onMomentumScrollEnd={e => onScrollSlide(e, setDotNumber, getPixel(360))}
        style={{width: width}}
        pagingEnabled
        horizontal>
        {Array.isArray(imageArray) &&
          imageArray.map((item, index) => {
            const image = typeof item === 'string' ? {uri: item} : item;
            return (
              <Image source={image} style={{width: width, height: height}} />
            );
          })}
        {(!Array.isArray(imageArray) || imageArray?.length < 1) && (
          <Image
            source={require('@assets/image/none_image_l.png')}
            style={{width, height}}
          />
        )}
      </ScrollView>
      <View style={styles.dotView}>
        {dotArray.map((item, index) => {
          return (
            <View
              style={{
                width: getPixel(7),
                height: getPixel(7),
                borderRadius: 100,
                backgroundColor:
                  index === dotNumber ? Theme.color.white : '#fff5',
                marginRight: index + 1 !== dotArray.length ? getPixel(6) : 0,
              }}
            />
          );
        })}
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  dotView: {
    position: 'absolute',
    bottom: getHeightPixel(12),
    left: 0,
    justifyContent: 'center',
    width: getPixel(360),
    flexDirection: 'row',
  },
  mainContainer: {flex: 1},
});
