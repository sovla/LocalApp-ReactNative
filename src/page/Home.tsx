import {
  Alert,
  GestureResponderEvent,
  StyleSheet,
  Text,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
} from 'react-native';
import React from 'react';

export interface HomeProp {
  title?: string;
  content?: string;
  onPressContent?(e: GestureResponderEvent): void;
}

export default function Home({title = 'Home', content = 'content', onPressContent}: HomeProp) {
  return (
    <View>
      <Text>{title}</Text>
      <TouchableOpacity testID="Button" onPress={onPressContent}>
        <Text>{content}</Text>
      </TouchableOpacity>
    </View>
  );
}

const styles = StyleSheet.create({});
