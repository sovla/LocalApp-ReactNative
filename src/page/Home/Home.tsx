import {View, ScrollView, Modal, TouchableOpacity, StyleSheet, NativeScrollEvent, NativeSyntheticEvent} from 'react-native';
import React, {useCallback, useLayoutEffect, useState} from 'react';
import Header from '@/Components/Home/Header';
import CategoryScroll from '@/Components/Home/CategoryScroll';
import HomeList from '@/Components/Home/HomeList';
import Theme from '@/assets/global/Theme';
import Footer from '@/Components/Home/Footer';
import Location from '@/Components/Modal/Location';
import ProductList from '@/Components/Home/ProductList';
import {getHeightPixel, getPixel} from '@/Util/pixelChange';
import AutoHeightImage from 'react-native-auto-height-image';
import useBoolean from '@/Hooks/useBoolean';
import UploadModal from '@/Components/Home/UploadModal';
import {useTranslation} from 'react-i18next';
import {HomeProps} from '@/Types/Screen/Screen';
import ModalPopup from '@/Components/Home/ModalPopup';
import {useFocusEffect, useIsFocused} from '@react-navigation/native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import useApi from '@/Hooks/useApi';
import {useAppSelector} from '@/Hooks/CustomHook';
import {HomeProductListType} from '@/Types/Components/HomeTypes';
export default function Home({navigation}: HomeProps): JSX.Element {
  const {t} = useTranslation();
  const fontSize = useAppSelector(state => state.fontSize.value);
  const {user} = useAppSelector(state => state);
  const [isList, setIsList] = useState(true);
  const {value: isUpload, on: onUpload, off: offUpload} = useBoolean(false);
  const {value: isPopup, on: onIsPopup, off: offIsPopup} = useBoolean(false);
  const {value: isChange, on: onIsChange, off: offIsChange} = useBoolean(false);
  const [page, setPage] = useState(1);
  const {data, isLoading, isError, errorMessage} = useApi<HomeProductListType['T'], HomeProductListType['D']>(
    null,
    'product_home_list.php',
    {
      mt_idx: user.mt_idx,
      page: page,
    },
    {
      focusRetry: true,
    },
  );

  const isFocused = useIsFocused();
  useLayoutEffect(() => {
    AsyncStorage.getItem('isPopup').then(result => {
      if (result === 'N') {
        onIsPopup();
      }
    });
  }, []);

  const onPressItem = useCallback((idx: string, cate: string) => {
    navigation.navigate('ProductDetail', {
      pt_cate: cate,
      pt_idx: idx,
    });
  }, []);

  const onScroll = useCallback((event: NativeSyntheticEvent<NativeScrollEvent>) => {
    if (event.nativeEvent.contentOffset.y > getHeightPixel(160)) {
      onIsChange();
    } else {
      offIsChange();
    }
  }, []);
  return (
    <View style={{flex: 1, backgroundColor: Theme.color.whiteGray_F6}}>
      <Header isChange={isChange} />
      <ScrollView onScroll={onScroll}>
        <CategoryScroll key="CategoryScroll" />
        <HomeList key="HomeList" location="Bom Retiro" isList={isList} setIsList={setIsList} />

        <ProductList isList={isList} list={data?.list ? data.list : []} onPressItem={onPressItem} />
      </ScrollView>

      <TouchableOpacity onPress={onUpload} style={styles.uploadTouch}>
        {!isUpload && <AutoHeightImage source={require('@assets/image/upload_blue.png')} width={getPixel(70)} />}
      </TouchableOpacity>
      <Footer menu="home" />
      {isUpload && (
        <Modal transparent visible>
          <UploadModal onClose={offUpload} />
        </Modal>
      )}
      {isPopup && <ModalPopup onClose={offIsPopup} />}
    </View>
  );
}

const styles = StyleSheet.create({
  uploadTouch: {
    position: 'absolute',
    right: getPixel(16),
    bottom: getHeightPixel(60),
  },
});
