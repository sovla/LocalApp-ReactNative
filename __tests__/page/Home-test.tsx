/**
 * @format
 */

import 'react-native';
import React from 'react';

import Home, {HomeProp} from '../../src/page/Home';
import {fireEvent, render} from '@testing-library/react-native';

function getHome(innerProps: HomeProp) {
  return <Home {...innerProps} />;
}

// jest 라이브러리 describe는 각각의 test들을 그룹핑(?) 해주고 모든 테스트는 test메서드로 할 수 있는데 test는 alias로 it이라는
//  메서드로도 사용할 수 있다.
describe('[Home] Render', () => {
  const props = {};
  const component = getHome(props);
  test('render without crashing', () => {
    const rendered = render(component);
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });
});

describe('[Home] Button Test', () => {
  const onPressMock = jest.fn();
  const props: HomeProp = {
    title: 'Hello!',
    content: 'Hello',
    onPressContent: onPressMock,
  };

  test('버튼이 눌린다.', () => {
    const rendered = render(<Home {...props} />);
    for (let i = 0; i < 5; i++) {
      fireEvent(rendered.getByTestId('Button'), 'onPress');
    }
    expect(onPressMock).toBeCalledTimes(5);

    expect(rendered.getByText('Hello'));
  });
});
