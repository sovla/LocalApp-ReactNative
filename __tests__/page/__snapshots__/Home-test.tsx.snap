// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Home] Render render without crashing 1`] = `
<View>
  <Text>
    Home
  </Text>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="Button"
  >
    <Text>
      content
    </Text>
  </View>
</View>
`;
